// byted-ailab-speech-sdk/dist/js/modern/features/tts/WebsocketMse.js
/* eslint-disable @typescript-eslint/no-use-before-define */

import { decodeAudioOnlyResponse, convertBlobToArrayBuffer, buildFullClientRequestBody } from ".";
export default function WebsocketMSE() {
  const context = {
    audioBuffers: [],
    sourceBuffer: undefined,
    socket: undefined,
    locked: false
  };

  /**
   * @param url 接入WebSocket的请求地址
   * @param config 接入tts配置
   * @param onStart WebSocket连接建立成功时的回调函数
   * @param onMessage WebSocket连接时的回调函数，返回合成音频文件ArrayBuffer
   * @param onClose WebSocket连接关闭时的回调函数
   * @param onError 服务出错的回调信息
   * @param onWSError WebSocket出错的回调信息
   */
  function start({
    url,
    config,
    onStart,
    onMessage,
    onClose,
    onError,
    onWSError,
    getSentences,
    debug = false
  }) {
    const mediaSource = new MediaSource();
    mediaSource.onsourceopen = () => {
      context.sourceBuffer = mediaSource.addSourceBuffer('audio/mpeg');
      context.sourceBuffer.onupdateend = () => addChunks();
    };
    const socket = new WebSocket(url);
    context.socket = socket;
    socket.onopen = () => {
      debug && console.info('connected');
      socket.send(buildFullClientRequestBody(config));
      onStart === null || onStart === void 0 || onStart();
    };
    socket.onmessage = async e => {
      if (e.data instanceof Blob) {
        const buffer = await convertBlobToArrayBuffer(e.data);
        decodeAudioOnlyResponse(buffer, (audioBuffer, finished, error) => {
          if (error) {
            onError === null || onError === void 0 || onError(error);
          }
          if (finished) {
            socket.close();
          }
          onMessage === null || onMessage === void 0 || onMessage(audioBuffer);
          context.audioBuffers.push(audioBuffer);
          addChunks();
        }, description => getSentences === null || getSentences === void 0 ? void 0 : getSentences(description));
      } else {
        try {
          const res = JSON.parse(e.data);
          if (res.errMsg) {
            onError === null || onError === void 0 || onError(res.errMsg);
          }
        } catch (error) {
          debug && console.error(error);
        }
      }
    };
    socket.onclose = () => {
      debug && console.info('disconnected');
      // clear audioBuffers
      const id = setInterval(() => {
        const isAllBlobProcessed = context.audioBuffers.length === 0;
        const isSourceBufferReady = context.sourceBuffer && !context.sourceBuffer.updating;
        if (isAllBlobProcessed && isSourceBufferReady) {
          try {
            mediaSource.endOfStream();
          } catch (e) {
            debug && console.error(e);
          }
          clearInterval(id);
        }
      }, 0);
      onClose === null || onClose === void 0 || onClose();
    };
    socket.onerror = err => onWSError === null || onWSError === void 0 ? void 0 : onWSError(err);
    return URL.createObjectURL(mediaSource);
  }
  function addChunks() {
    const {
      sourceBuffer,
      audioBuffers
    } = context;
    if (!context.locked) {
      context.locked = true;
      if (audioBuffers.length && sourceBuffer && !sourceBuffer.updating) {
        try {
          sourceBuffer.appendBuffer(audioBuffers.shift());
        } catch (e) {
          console.error(e);
        }
      }
      context.locked = false;
    }
  }

  /**
   * 手动关闭ws链接
   */
  function close() {
    var _context$socket;
    (_context$socket = context.socket) === null || _context$socket === void 0 || _context$socket.close();
  }
  return Object.freeze({
    start,
    close
  });
}
