// byted-ailab-speech-sdk/dist/js/modern/features/bidirectional-tts/constants.js
export let PROTOCOL_CONSTANTS = /*#__PURE__*/function (PROTOCOL_CONSTANTS) {
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["PROTOCOL_VERSION"] = 1] = "PROTOCOL_VERSION";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["DEFAULT_HEADER_SIZE"] = 1] = "DEFAULT_HEADER_SIZE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["PROTOCOL_VERSION_BITS"] = 4] = "PROTOCOL_VERSION_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["HEADER_BITS"] = 4] = "HEADER_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["MESSAGE_TYPE_BITS"] = 4] = "MESSAGE_TYPE_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["MESSAGE_TYPE_SPECIFIC_FLAGS_BITS"] = 4] = "MESSAGE_TYPE_SPECIFIC_FLAGS_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["MESSAGE_SERIALIZATION_BITS"] = 4] = "MESSAGE_SERIALIZATION_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["MESSAGE_COMPRESSION_BITS"] = 4] = "MESSAGE_COMPRESSION_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["RESERVED_BITS"] = 8] = "RESERVED_BITS";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["EVENT_NUMBER_BYTES"] = 4] = "EVENT_NUMBER_BYTES";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["PAYLOAD_SIZE_BYTES"] = 4] = "PAYLOAD_SIZE_BYTES";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["CLIENT_FULL_REQUEST"] = 1] = "CLIENT_FULL_REQUEST";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["CLIENT_AUDIO_ONLY_REQUEST"] = 2] = "CLIENT_AUDIO_ONLY_REQUEST";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["SERVER_FULL_RESPONSE"] = 9] = "SERVER_FULL_RESPONSE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["SERVER_ACK"] = 11] = "SERVER_ACK";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["SERVER_ERROR_RESPONSE"] = 15] = "SERVER_ERROR_RESPONSE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["NO_SEQUENCE"] = 0] = "NO_SEQUENCE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["POS_SEQUENCE"] = 1] = "POS_SEQUENCE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["NEG_WITHOUT_SEQUENCE"] = 2] = "NEG_WITHOUT_SEQUENCE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["NEG_WITH_SEQUENCE"] = 3] = "NEG_WITH_SEQUENCE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["MESSAGE_WITH_EVENT"] = 4] = "MESSAGE_WITH_EVENT";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["NO_SERIALIZATION"] = 0] = "NO_SERIALIZATION";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["JSON"] = 1] = "JSON";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["THRIFT"] = 3] = "THRIFT";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["CUSTOM_TYPE"] = 15] = "CUSTOM_TYPE";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["NO_COMPRESSION"] = 0] = "NO_COMPRESSION";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["GZIP"] = 1] = "GZIP";
  PROTOCOL_CONSTANTS[PROTOCOL_CONSTANTS["CUSTOM_COMPRESSION"] = 15] = "CUSTOM_COMPRESSION";
  return PROTOCOL_CONSTANTS;
}({});
export let MESSAGE_TYPE_ENUM = /*#__PURE__*/function (MESSAGE_TYPE_ENUM) {
  MESSAGE_TYPE_ENUM[MESSAGE_TYPE_ENUM["CLIENT_FULL_REQUEST"] = 1] = "CLIENT_FULL_REQUEST";
  MESSAGE_TYPE_ENUM[MESSAGE_TYPE_ENUM["CLIENT_AUDIO_ONLY_REQUEST"] = 2] = "CLIENT_AUDIO_ONLY_REQUEST";
  MESSAGE_TYPE_ENUM[MESSAGE_TYPE_ENUM["SERVER_FULL_RESPONSE"] = 9] = "SERVER_FULL_RESPONSE";
  MESSAGE_TYPE_ENUM[MESSAGE_TYPE_ENUM["SERVER_AUDIO_ONLY_RESPONSE"] = 11] = "SERVER_AUDIO_ONLY_RESPONSE";
  MESSAGE_TYPE_ENUM[MESSAGE_TYPE_ENUM["SERVER_ERROR_RESPONSE"] = 15] = "SERVER_ERROR_RESPONSE";
  return MESSAGE_TYPE_ENUM;
}({});
export let MESSAGE_FLAG_ENUM = /*#__PURE__*/function (MESSAGE_FLAG_ENUM) {
  MESSAGE_FLAG_ENUM[MESSAGE_FLAG_ENUM["NO_SEQUENCE"] = 0] = "NO_SEQUENCE";
  MESSAGE_FLAG_ENUM[MESSAGE_FLAG_ENUM["POS_SEQUENCE"] = 1] = "POS_SEQUENCE";
  MESSAGE_FLAG_ENUM[MESSAGE_FLAG_ENUM["NEG_WITHOUT_SEQUENCE"] = 2] = "NEG_WITHOUT_SEQUENCE";
  MESSAGE_FLAG_ENUM[MESSAGE_FLAG_ENUM["NEG_WITH_SEQUENCE"] = 3] = "NEG_WITH_SEQUENCE";
  MESSAGE_FLAG_ENUM[MESSAGE_FLAG_ENUM["MESSAGE_WITH_EVENT"] = 4] = "MESSAGE_WITH_EVENT";
  return MESSAGE_FLAG_ENUM;
}({}); // 带event的消息
export let SerializationEnum = /*#__PURE__*/function (SerializationEnum) {
  SerializationEnum[SerializationEnum["NO_SERIALIZATION"] = 0] = "NO_SERIALIZATION";
  SerializationEnum[SerializationEnum["JSON"] = 1] = "JSON";
  SerializationEnum[SerializationEnum["THRIFT"] = 3] = "THRIFT";
  SerializationEnum[SerializationEnum["CUSTOM_TYPE"] = 15] = "CUSTOM_TYPE";
  return SerializationEnum;
}({});
export let CompressionEnum = /*#__PURE__*/function (CompressionEnum) {
  CompressionEnum[CompressionEnum["NO_COMPRESSION"] = 0] = "NO_COMPRESSION";
  CompressionEnum[CompressionEnum["GZIP"] = 1] = "GZIP";
  CompressionEnum[CompressionEnum["CUSTOM_COMPRESSION"] = 15] = "CUSTOM_COMPRESSION";
  return CompressionEnum;
}({});
export let CLIENT_EVENT_ENUM = /*#__PURE__*/function (CLIENT_EVENT_ENUM) {
  CLIENT_EVENT_ENUM[CLIENT_EVENT_ENUM["StartConnection"] = 1] = "StartConnection";
  CLIENT_EVENT_ENUM[CLIENT_EVENT_ENUM["FinishConnection"] = 2] = "FinishConnection";
  CLIENT_EVENT_ENUM[CLIENT_EVENT_ENUM["StartSession"] = 100] = "StartSession";
  CLIENT_EVENT_ENUM[CLIENT_EVENT_ENUM["FinishSession"] = 102] = "FinishSession";
  CLIENT_EVENT_ENUM[CLIENT_EVENT_ENUM["TaskRequest"] = 200] = "TaskRequest";
  return CLIENT_EVENT_ENUM;
}({}); // 任务请求
export let SERVER_EVENT_ENUM = /*#__PURE__*/function (SERVER_EVENT_ENUM) {
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["ConnectionStarted"] = 50] = "ConnectionStarted";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["ConnectionFailed"] = 51] = "ConnectionFailed";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["ConnectionFinished"] = 52] = "ConnectionFinished";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["SessionStarted"] = 150] = "SessionStarted";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["SessionCanceled"] = 151] = "SessionCanceled";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["SessionFinished"] = 152] = "SessionFinished";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["SessionFailed"] = 153] = "SessionFailed";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["TTSSentenceStart"] = 350] = "TTSSentenceStart";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["TTSSentenceEnd"] = 351] = "TTSSentenceEnd";
  SERVER_EVENT_ENUM[SERVER_EVENT_ENUM["TTSResponse"] = 352] = "TTSResponse";
  return SERVER_EVENT_ENUM;
}({}); // 返回句子的音频内容
