// byted-ailab-speech-sdk/dist/js/treeshaking/features/bidirectional-tts/WebsocketMse.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _merge from "lodash-es/merge";
import _cloneDeep from "lodash-es/cloneDeep";
/* eslint-disable @typescript-eslint/no-use-before-define */
import { arrayBufferToString } from "../../utils/arraybuffer-to-string";
import { stringToArraybuffer } from "../../utils/string-to-arraybuffer";
import { uuid } from "../../utils/uuid";
import { CLIENT_EVENT_ENUM, MESSAGE_TYPE_ENUM, PROTOCOL_CONSTANTS, SERVER_EVENT_ENUM } from "./constants";
import { buildFullClientRequestBody, getEventName } from "./utils";
import { safeJsonParse } from "../../utils/safe-json-parse";
export default function WebsocketMSE() {
  const context = {
    audioBuffers: [],
    sourceBuffer: undefined,
    socket: undefined,
    locked: false,
    sessionId: '',
    connectId: '',
    sessionConfig: {},
    sessionStarted: false
  };
  function getSessionOffset() {
    return context.sessionId ? 4 + stringToArraybuffer(context.sessionId).byteLength : 0;
  }

  /**
   * @param url 接入WebSocket的请求地址
   * @param config 接入tts配置
   * @param onStart WebSocket连接建立成功时的回调函数
   * @param onMessage WebSocket连接时的回调函数，返回合成音频文件ArrayBuffer
   * @param onClose WebSocket连接关闭时的回调函数
   * @param onError 服务出错的回调信息
   * @param onWSError WebSocket出错的回调信息
   */
  function start({
    url,
    config,
    onStart,
    onMessage,
    onClose,
    onError,
    onWSError,
    getSentences,
    onSessionStarted,
    onTTSSentenceStart,
    onTTSSentenceEnd,
    debug = false
  }) {
    const mediaSource = new MediaSource();
    mediaSource.onsourceopen = () => {
      context.sourceBuffer = mediaSource.addSourceBuffer('audio/mpeg');
      context.sourceBuffer.onupdateend = () => addChunks();
    };
    const socket = new WebSocket(url);
    context.socket = socket;
    socket.onopen = () => {
      debug && console.info('connected');
      socket.send(buildFullClientRequestBody(CLIENT_EVENT_ENUM.StartConnection));
      onStart === null || onStart === void 0 || onStart();
    };
    socket.onmessage = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(function* (e) {
        // 解析二进制帧
        if (e.data instanceof Blob) {
          // 定义偏移量
          let offset = 0;
          const respBuffer = yield e.data.arrayBuffer();
          const dv = new DataView(respBuffer);
          const version = dv.getUint8(0) >> 4;
          const header = dv.getUint8(0) & 0x0f;
          const messageType = dv.getUint8(1) >> 4;
          const messageFlag = dv.getUint8(1) & 0x0f;
          const serializationMethod = dv.getUint8(2) >> 4;
          const compression = dv.getUint8(2) & 0x0f;
          const reserved = dv.getUint8(3);
          offset += 4;
          // console.log('version:', version.toString(2).padStart(4, '0'));
          // console.log('header:', header.toString(2).padStart(4, '0'));
          // console.log('messageType:', messageType.toString(2).padStart(4, '0'));
          // console.log('messageFlag:', messageFlag.toString(2).padStart(4, '0'));
          // console.log('serializationMethod:', serializationMethod.toString(2).padStart(4, '0'));
          // console.log('compression:', compression.toString(2).padStart(4, '0'));
          // console.log('reserved:', reserved.toString(2).padStart(4, '0'));

          // console.log('========服务头信息========');

          // console.log('头长度:', header);
          // console.log('消息类型:', MESSAGE_TYPE_ENUM[messageType]);
          // console.log('消息标识:', MESSAGE_FLAG_ENUM[messageFlag]);
          // console.log('序列化:', SerializationEnum[serializationMethod]);
          // console.log('压缩:', CompressionEnum[compression]);
          if (messageType === MESSAGE_TYPE_ENUM.SERVER_FULL_RESPONSE || messageType === MESSAGE_TYPE_ENUM.SERVER_AUDIO_ONLY_RESPONSE) {
            if (messageFlag === PROTOCOL_CONSTANTS.MESSAGE_WITH_EVENT) {
              // 四个字节获取事件number
              const eventNumber = dv.getUint32(offset);
              offset += 4;
              const eventName = getEventName(eventNumber);
              console.log('server eventName:', eventName);

              // 无需解析
              if (eventNumber === SERVER_EVENT_ENUM.ConnectionStarted) {
                // 收到ConnectionStarted事件后，自动发送StartSession事件
                const sessionId = uuid();
                context.sessionId = sessionId;
                context.sessionConfig = _cloneDeep(config);
                socket.send(buildFullClientRequestBody(CLIENT_EVENT_ENUM.StartSession, {
                  payload: context.sessionConfig,
                  session_id: context.sessionId
                }));
                // 无需解析  SessionStarted
              } else if (eventNumber === SERVER_EVENT_ENUM.SessionStarted) {
                context.sessionStarted = true;
                onSessionStarted === null || onSessionStarted === void 0 || onSessionStarted();
              } else if (eventNumber === SERVER_EVENT_ENUM.TTSResponse) {
                if (messageType === MESSAGE_TYPE_ENUM.SERVER_AUDIO_ONLY_RESPONSE) {
                  const audioSegment = respBuffer.slice(offset + getSessionOffset());
                  const audioLength = new DataView(audioSegment).getUint32(0);
                  // console.log('audioLength:', audioLength);
                  const audioBuffer = audioSegment.slice(4);
                  // console.log('audioBuffer:', audioBuffer);
                  context.audioBuffers.push(audioBuffer);
                  addChunks();
                  onMessage === null || onMessage === void 0 || onMessage(audioBuffer);
                } else {
                  console.log('不支持的TTSResponse');
                }
              } else if (eventNumber === SERVER_EVENT_ENUM.TTSSentenceStart) {
                const payloadBuffer = respBuffer.slice(offset + getSessionOffset() + 4);
                const payload = arrayBufferToString(payloadBuffer);
                onTTSSentenceStart === null || onTTSSentenceStart === void 0 || onTTSSentenceStart(safeJsonParse(payload));
              } else if (eventNumber === SERVER_EVENT_ENUM.TTSSentenceEnd) {
                const payloadBuffer = respBuffer.slice(offset + getSessionOffset() + 4);
                const payload = arrayBufferToString(payloadBuffer);
                onTTSSentenceEnd === null || onTTSSentenceEnd === void 0 || onTTSSentenceEnd(safeJsonParse(payload));
              } else if ([SERVER_EVENT_ENUM.SessionFinished, SERVER_EVENT_ENUM.SessionCanceled, SERVER_EVENT_ENUM.SessionFailed].includes(eventNumber)) {
                try {
                  const payloadBuffer = respBuffer.slice(offset + getSessionOffset() + 4);
                  const str = arrayBufferToString(payloadBuffer);
                  const {
                    status_code,
                    message
                  } = JSON.parse(str);
                  if (status_code !== 20000000) {
                    onError === null || onError === void 0 || onError({
                      msg: message,
                      code: status_code
                    });
                  }
                } catch (err) {
                  console.log('Session error:', err);
                }
              } else if (eventNumber === SERVER_EVENT_ENUM.ConnectionFailed) {
                const payloadBuffer = respBuffer.slice(offset + 8);
                try {
                  const str = arrayBufferToString(payloadBuffer);
                  const {
                    status_code,
                    message
                  } = JSON.parse(str);
                  if (status_code !== 20000000) {
                    onError === null || onError === void 0 || onError({
                      msg: message,
                      code: status_code
                    });
                  }
                } catch (err) {
                  console.log('ConnectionFailed error:', err);
                }
              }
            } else {
              console.log('unable to handle unknown messageFlag', messageFlag);
            }
          } else if (messageType === MESSAGE_TYPE_ENUM.SERVER_ERROR_RESPONSE) {
            try {
              const errorSegment = respBuffer.slice(offset + getSessionOffset());
              // const errorLength = new DataView(errorSegment).getUint32(0);
              offset += 4;
              const errorBuffer = errorSegment.slice(4);
              const str = arrayBufferToString(errorBuffer);
              const {
                status_code,
                message
              } = JSON.parse(str);
              onError === null || onError === void 0 || onError({
                msg: message,
                code: status_code
              });
            } catch (err) {
              console.log('server error response:', err);
            }
          } else {
            console.log('unable to handle messageType:', messageType);
          }
        }
      });
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }();
    socket.onclose = () => {
      debug && console.info('disconnected');
      // clear audioBuffers
      const id = setInterval(() => {
        const isAllBlobProcessed = context.audioBuffers.length === 0;
        const isSourceBufferReady = context.sourceBuffer && !context.sourceBuffer.updating;
        if (isAllBlobProcessed && isSourceBufferReady) {
          try {
            mediaSource.endOfStream();
          } catch (e) {
            debug && console.error(e);
          }
          clearInterval(id);
        }
      }, 0);
      onClose === null || onClose === void 0 || onClose();
    };
    socket.onerror = err => onWSError === null || onWSError === void 0 ? void 0 : onWSError(err);
    return URL.createObjectURL(mediaSource);
  }
  function addChunks() {
    const {
      sourceBuffer,
      audioBuffers
    } = context;
    if (!context.locked) {
      context.locked = true;
      if (audioBuffers.length && sourceBuffer && !sourceBuffer.updating) {
        try {
          sourceBuffer.appendBuffer(audioBuffers.shift());
        } catch (e) {
          console.error(e);
        }
      }
      context.locked = false;
    }
  }

  /**
   * 客户端关闭ws链接
   */
  function close() {
    var _context$socket;
    (_context$socket = context.socket) === null || _context$socket === void 0 || _context$socket.close();
  }

  /**
   * 发送文本
   */
  function sendText(text) {
    var _context$socket2;
    if (!context.sessionStarted) {
      return;
    }
    const config = _merge(context.sessionConfig, {
      req_params: {
        text
      }
    });
    (_context$socket2 = context.socket) === null || _context$socket2 === void 0 || _context$socket2.send(buildFullClientRequestBody(CLIENT_EVENT_ENUM.TaskRequest, {
      payload: config,
      session_id: context.sessionId
    }));
  }

  /**
   * 结束一个session
   */
  function finishSession() {
    var _context$socket3;
    (_context$socket3 = context.socket) === null || _context$socket3 === void 0 || _context$socket3.send(buildFullClientRequestBody(CLIENT_EVENT_ENUM.FinishSession, {
      session_id: context.sessionId
    }));
  }

  /**
   * 发送一个断联
   */
  function finishConnection() {
    var _context$socket4;
    (_context$socket4 = context.socket) === null || _context$socket4 === void 0 || _context$socket4.send(buildFullClientRequestBody(CLIENT_EVENT_ENUM.FinishConnection));
  }
  return Object.freeze({
    start,
    close,
    sendText,
    finishSession,
    finishConnection
  });
}
