// byted-ailab-speech-sdk/dist/js/node/features/asr/websocket-record-rtc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebsocketRecordRtc = WebsocketRecordRtc;
var _get = _interopRequireDefault(require("lodash-es/get"));
var _recordrtc = _interopRequireWildcard(require("recordrtc"));
var _utils = require("./utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function WebsocketRecordRtc(params = {}) {
  const context = _objectSpread(_objectSpread({}, params), {}, {
    webSocket: undefined,
    mediaStream: undefined,
    recorder: undefined,
    streamArr: undefined
  });
  /**
   *
   * @param url 服务端提供Socket服务的URL接口
   * @param postData Socket建立连接时发送到服务端的数据
   * @param onStart Socket连接建立成功时的回调函数
   * @param onMessage Socket包到达时的处理函数
   * @param onClose Socket连接关闭时的回调函数
   */
  function connect({
    url,
    config,
    debug = false
  }) {
    const {
      onStart,
      onMessage,
      onClose,
      onError
    } = context;
    const socket = new WebSocket(url);
    context.webSocket = socket;
    socket.onopen = () => {
      debug && console.info('socket connected');
      const build = (0, _utils.encodeFullClientRequest)(config);
      socket.send(build);
      onStart === null || onStart === void 0 || onStart();
    };
    socket.onmessage = async event => {
      try {
        const isbigasr = (0, _get.default)(config, 'request.model_name') === 'bigmodel';
        const res_json = await (0, _utils.parseResponse)(event.data, {
          hasSequence: isbigasr
        });
        if (!isbigasr) {
          const text = res_json.result ? res_json.result[0].text : '';
          onMessage === null || onMessage === void 0 || onMessage(text, res_json);
        } else {
          const text = res_json.result ? res_json.result.text : '';
          onMessage === null || onMessage === void 0 || onMessage(text, res_json);
        }
      } catch (error) {
        console.error(error);
      }
    };
    socket.onclose = () => {
      debug && console.info('socket onclose disconnected');
      onClose === null || onClose === void 0 || onClose();
    };
    socket.onerror = () => {
      debug && console.info('socket onerror disconnected');
      onError === null || onError === void 0 || onError();
    };
  }
  async function startRecord(options, handleRecordResult) {
    context.mediaStream = await window.navigator.mediaDevices.getUserMedia({
      audio: true
    });
    if (!context.streamArr) {
      context.streamArr = [context.mediaStream];
    } else {
      context.streamArr.push(context.mediaStream);
    }
    context.recorder = new _recordrtc.default(context.mediaStream, _objectSpread(_objectSpread({
      // 后端要求单通道，16K采样率，PCM
      type: 'audio',
      recorderType: _recordrtc.StereoAudioRecorder,
      mimeType: 'audio/wav',
      numberOfAudioChannels: 1,
      desiredSampRate: 16000,
      disableLogs: true,
      timeSlice: 200,
      bufferSize: 4096
    }, options), {}, {
      ondataavailable(recordResult) {
        const socket = context.webSocket;
        if (socket) {
          handleRecordResult === null || handleRecordResult === void 0 || handleRecordResult(recordResult);
          const pcm = recordResult.slice(44);
          const data = (0, _utils.encodeAudioOnlyRequest)(pcm);
          if (socket.readyState === socket.OPEN) {
            socket.send(data);
          }
        }
      }
    }));
    context.recorder.startRecording();
  }
  function stopRecord() {
    if (!context.recorder) {
      return;
    }
    context.recorder.stopRecording(() => {
      var _context$streamArr;
      if (!context.webSocket) {
        return;
      }
      context.webSocket.close();
      (_context$streamArr = context.streamArr) === null || _context$streamArr === void 0 || _context$streamArr.forEach(item => item.getAudioTracks().forEach(track => track.stop()));
      context.streamArr = undefined;
    });
  }
  return Object.freeze({
    connect,
    startRecord,
    stopRecord
  });
}
